kubernetes/updater/pkg/controller/deployment.go

```diff
@@ -0,0 +1,140 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+	"github.com/gravitational/trace"
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// DeploymentVersionUpdater Reconciles a deployment by changing its image
+type DeploymentVersionUpdater struct {
+	VersionUpdater
+	kclient.Client
+	Scheme *runtime.Scheme
+}
+
+// Reconcile treats a reconciliation request for a Deployment object. It gets the
+// object, retrieves it current version, and call the VersionUpdater to find
+// if the Deployment should be updated. If it's the case, it changes the
+// Teleport image version and updates the Deployment in Kubernetes.
+func (r *DeploymentVersionUpdater) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx).WithValues("namespacedname", req.NamespacedName)
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	// Get the object we are reconciling
+	var obj appsv1.Deployment
+	if err := r.Get(ctx, req.NamespacedName, &obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			// log.Info("not found")
+			return ctrl.Result{}, nil
+		}
+		// log.Error(err, "failed to get resource")
+		return ctrl.Result{}, trace.Wrap(err)
+	}
+
+	// Get the current and past version
+	currentVersion, err := getDeploymentVersion(&obj)
+	if err != nil {
+		switch trace.Unwrap(err).(type) {
+		case *trace.BadParameterError:
+			log.Info("Teleport container found, but failed to get version from the img tag. Will continue and do a version update.")
+		default:
+			return requeueLater, trace.Wrap(err)
+		}
+	}
+
+	image, err := r.Do(ctx, &obj, currentVersion)
+	var (
+		noNewVersionErr *NoNewVersionError
+		maintenanceErr  *MaintenanceNotTriggeredError
+	)
+	switch {
+	case errors.As(err, &noNewVersionErr):
+		// Error contains the detected versions
+		log.Info("Deployment is already up-to-date, not updating.", "err", err)
+		return requeueLater, nil
+	case errors.As(err, &maintenanceErr):
+		// Not logging the error because it provides no other information than its type.
+		log.Info("No maintenance triggered, not updating.", "currentVersion", currentVersion)
+		return requeueLater, nil
+	case trace.IsTrustError(err):
+		// Logging as error as image verification should not fail under normal use
+		log.Error(err, "Image verification failed, not updating.")
+		return requeueLater, trace.Wrap(err)
+	case err != nil:
+		log.Error(err, "Unexpected error, not updating.")
```

Comment:
Will this give us stack trace? We usually do `log.WithError(err).Error("xxx")` but we use logrus in Teleport, not sure if this logger supports it too.

Reply:
I checked and it doesn't. We are using the `logr` interface backed by `zap`. It shows the stack trace up to the `log.Error` but doesn't unpack all errors as we do in teleport.

Reply:
I played a bit with the logger implementation and it does not seem to be possible to print pretty traces without doing our own formatting. Can I ignore this for now as this issue also exists for the operator?

If we decide this is worth the time I would suggest adding logr/zap support in `gravitaional/trace` so the operator and updater can benefit from improved error logging.